File: solutions/cpp/Build_BST_from_sorted_array.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <random>
#include <vector>

#include "./BST_prototype.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::unique_ptr;
using std::vector;

unique_ptr<BSTNode<int>> BuildMinHeightBSTFromSortedArrayHelper(
    const vector<int>&, int, int);

// @include
unique_ptr<BSTNode<int>> BuildMinHeightBSTFromSortedArray(
    const vector<int>& A) {
  return BuildMinHeightBSTFromSortedArrayHelper(A, 0, A.size());
}

// Build a min-height BST over the entries in A[start : end - 1].
unique_ptr<BSTNode<int>> BuildMinHeightBSTFromSortedArrayHelper(
    const vector<int>& A, int start, int end) {
  if (start >= end) {
    return nullptr;
  }
  int mid = start + ((end - start) / 2);
  return unique_ptr<BSTNode<int>>(new BSTNode<int>{A[mid],
      BuildMinHeightBSTFromSortedArrayHelper(A, start, mid),
      BuildMinHeightBSTFromSortedArrayHelper(A, mid + 1, end)});
}
// @exclude

template <typename T>
void TraversalCheck(const unique_ptr<BSTNode<T>>& tree, T* target) {
  if (tree) {
    TraversalCheck(tree->left, target);
    assert(*target == tree->data);
    ++*target;
    TraversalCheck(tree->right, target);
  }
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    vector<int> A;
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 1000);
      n = dis(gen);
    }
    for (int i = 0; i < n; ++i) {
      A.emplace_back(i);
    }
    unique_ptr<BSTNode<int>> tree = BuildMinHeightBSTFromSortedArray(A);
    int target = 0;
    TraversalCheck<int>(tree, &target);
  }
  return 0;
}

File: static/cpp/java/src/main/java/com/epi/BuildBSTFromSortedArray.java
package com.epi;

import com.epi.BinarySearchTreePrototypeTemplate.BSTNode;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class BuildBSTFromSortedArray {
  // @include
  public static BSTNode<Integer> buildMinHeightBSTFromSortedArray(
      List<Integer> A) {
    return buildMinHeightBSTFromSortedArrayHelper(A, 0, A.size());
  }

  // Build a min-height BST over the entries in A.subList(start, end - 1).
  private static BSTNode<Integer> buildMinHeightBSTFromSortedArrayHelper(
      List<Integer> A, int start, int end) {
    if (start >= end) {
      return null;
    }
    int mid = start + ((end - start) / 2);
    return new BSTNode<>(
        A.get(mid), buildMinHeightBSTFromSortedArrayHelper(A, start, mid),
        buildMinHeightBSTFromSortedArrayHelper(A, mid + 1, end));
  }
  // @exclude

  private static int traversalCheck(BSTNode<Integer> tree, Integer target) {
    if (tree != null) {
      target = traversalCheck(tree.left, target);
      assert(target.equals(tree.data));
      ++target;
      target = traversalCheck(tree.right, target);
    }
    return target;
  }

  private static void simpleTest() {
    BSTNode<Integer> result
        = buildMinHeightBSTFromSortedArray(Arrays.asList(1, 2, 3, 4));
    assert(3 == result.data);
    assert(2 == result.left.data);
    assert(1 == result.left.left.data);
    assert(4 == result.right.data);
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = r.nextInt(1000) + 1;
      }
      List<Integer> A = new ArrayList<>(n);
      for (int i = 0; i < n; ++i) {
        A.add(i);
      }
      BSTNode<Integer> tree = buildMinHeightBSTFromSortedArray(A);
      int target = 0;
      traversalCheck(tree, target);
    }
  }
}
